services/monitor/src/PendingContract.ts

```diff
@@ -0,0 +1,165 @@
+import logger from "./logger";
+import { FileHash } from "./util";
+import { Metadata, MetadataSourceMap } from "@ethereum-sourcify/lib-sourcify";
+import { id as keccak256str } from "ethers";
+import { KnownDecentralizedStorageFetchers } from "./types";
+import assert from "assert";
+import dotenv from "dotenv";
+dotenv.config();
+
+export default class PendingContract {
+  public metadataHash: FileHash;
+  public address: string;
+  public chainId: number;
+  public metadata: Metadata | undefined;
+  public pendingSources: MetadataSourceMap = {};
+  public fetchedSources: MetadataSourceMap = {};
+  private decentralizedStorageFetchers: KnownDecentralizedStorageFetchers;
+
+  constructor(
+    metadataHash: FileHash,
+    address: string,
+    chainId: number,
+    decentralizedStorageFetchers: KnownDecentralizedStorageFetchers
+  ) {
+    this.metadataHash = metadataHash;
+    this.address = address;
+    this.chainId = chainId;
+    this.decentralizedStorageFetchers = decentralizedStorageFetchers;
+  }
+
+  /**
+   * Assembles this contract by first fetching its metadata and then fetching all the sources listed in the metadata.
+   */
+  assemble = async () => {
+    const metadataFetcher =
+      this.decentralizedStorageFetchers[this.metadataHash.origin];
+    if (!metadataFetcher) {
+      throw new Error(
+        `No metadata fetcher found for origin ${this.metadataHash.origin}`
+      );
+    }
+    const metadataStr = await metadataFetcher
+      .fetch(this.metadataHash)
+      .catch((err) => {
+        throw new Error(
+          `Can't fetch metadata for ${this.address} ${this.metadataHash.hash}: ${err}`
+        );
+      });
+    logger.info(
+      `Fetched metadata for ${this.address} on chain ${this.chainId} from ${this.metadataHash.origin}`
+    );
+    // TODO: check if metadata hash matches this.metadataHash.hash
+    this.metadata = JSON.parse(metadataStr) as Metadata;
+    this.pendingSources = { ...this.metadata.sources }; // Copy, don't mutate original.
```

Comment:
Be aware that { ...object} clones only the first level properties, but deep properties are "cloned" as references to the first object. The only way to safely deep clone an object is to use `structuredClone`

Reply:
Good to know thanks